app/discovery/provider/consulcatalog/consulcatalog.go

```diff
@@ -0,0 +1,197 @@
+package consulcatalog
+
+import (
+	"context"
+	"fmt"
+	"github.com/umputun/reproxy/app/discovery"
+	"log"
+	"regexp"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+)
+
+//go:generate moq -out consul_client_mock.go -skip-ensure -fmt goimports . ConsulClient
+
+// ConsulClient defines interface getting consul services
+type ConsulClient interface {
+	Get() ([]consulService, error)
+}
+
+// ConsulCatalog provider periodically gets consul services with tags, started with 'reproxy.'
+// It stores service list IDs in the internal storage. If service list was changed, it send signal to the core
+// The provider maps services with rules, described in the docker provider documentation
+// Blockquote:
+// maps by default from ^/api/%s/(.*) to http://%s:%d/$1, i.e. http://example.com/api/my_container/something
+// will be mapped to http://172.17.42.1:8080/something. Ip will be the internal ip of the container and port exposed
+// in the Dockerfile.
+// Alternatively labels can alter this. reproxy.route sets source route, and reproxy.dest sets the destination.
+// Optional reproxy.server enforces match by server name (hostname) and reproxy.ping sets the health check url
+type ConsulCatalog struct {
+	client          ConsulClient
+	autoAPI         bool
+	refreshInterval time.Duration
+	listMx          sync.RWMutex
+	// current services list with ServiceID as map key
+	list map[string]consulService
+}
+
+// New creates new ConsulCatalog instance
+func New(client ConsulClient, autoAPI bool, refreshInterval time.Duration) *ConsulCatalog {
+	cc := &ConsulCatalog{
+		client:          client,
+		autoAPI:         autoAPI,
```

Comment:
I asked in prev review - what autoAPI even means here? In case of docker the goal is get all containers with exposed ports and it has the logic to restict this selection. Nothing like this I can see here.